### ISSUE TITLE
Build fails on AArch64 when configured with "configure.py --cpu=arm32". Error: unrecognized command line option ‘-mfpu=neon’

### ISSUE THREAD

**Issue by DevSidious** [randombit/botan#1442]
```
INFO: ./configure.py invoked with options "--cpu=arm32 --minimized-build --enable-modules=x509,ecdsa,rsa,emsa_pkcs1,sha2_64,aes --disable-shared-library --amalgamation --single-amalgamation-file"
INFO: Guessing target OS is linux (use --os to set)
INFO: Guessing to use compiler gcc (use --cc or CXX to set)
INFO: Auto-detected compiler version 5.4
INFO: Target is gcc:5.4-linux-arm32
INFO: Skipping (incompatible CPU): aes_ni aes_ssse3 chacha_sse2 clmul clmul_ssse3 idea_sse2 rdrand_rng rdseed sha1_sse2 sha1_x86 sha2_32_x86 shacal2_x86 threefish_avx2
INFO: Skipping (incompatible OS): darwin_secrandom getentropy win32_stats
INFO: Skipping (incompatible compiler): aes_armv8 pmull sha1_armv8 sha2_32_armv8
INFO: Skipping (not requested): adler32 aead aont aria auto_rng bcrypt blake2 blowfish camellia cascade cast cbc cbc_mac ccm cecpq1 certstor_sql certstor_sqlite3 cfb chacha chacha20poly1305 chacha_rng checksum cmac codec_filt comb4p compression crc24 crc32 cryptobox ctr curve25519 des dev_random dh dl_algo dl_group dlies dsa dyn_load eax ecdh ecgdsa ecies eckcdsa ed25519 elgamal eme_oaep eme_pkcs1 eme_raw emsa_raw emsa_x931 fd_unix ffi filters fpe_fe1 gcm gmac gost_28147 gost_3410 gost_3411 hkdf hmac hmac_drbg hotp http_util idea iso9796 kasumi kdf1 kdf1_iso18033 kdf2 keccak lion locking_allocator mce mceies md4 md5 misty1 mode_pad modes newhope nist_keywrap noekeon noekeon_simd ocb ofb par_hash passhash9 pbes2 pbkdf pbkdf1 pbkdf2 pgp_s2k pkcs11 poly1305 poly_dbl prf_tls prf_x942 proc_walk psk_db rc4 rdrand rfc3394 rfc6979 rmd160 salsa20 seed serpent serpent_simd sessions_sql sessions_sqlite3 sha3 shacal2 shacal2_simd shake shake_cipher simd siphash siv skein sm2 sm3 sm4 socket sp800_108 sp800_56a sp800_56c srp6 stateful_rng stream streebog system_rng thread_utils threefish tiger tls tls_cbc tss twofish whirlpool x919_mac xmss xtea xts
INFO: Skipping (requires external dependency): bearssl boost bzip2 lzma openssl sqlite3 tpm zlib
INFO: Loading modules: aes asn1 base base64 bigint block cpuid ec_gfp ec_group ecc_key ecdsa emsa1 emsa_pkcs1 emsa_pssr entropy hash hash_id hex kdf keypair mac mdx_hash mgf1 mp numbertheory pem pk_pad pubkey rng rsa sha1 sha2_32 sha2_64 utils x509
INFO: Using symlink to link files into build dir (use --link-method to change)
INFO: Writing amalgamation header to botan_all.h
INFO: Writing amalgamation header to botan_all_internal.h
INFO: Writing amalgamation source to botan_all.cpp
INFO: Botan 2.5.0 (revision git:0d63e98483e304a385013a4568312ab94dff3822) (unreleased undated) build setup is complete

make
...
...
...
g++ -fstack-protector -pthread -mfpu=neon -std=c++11 -D_REENTRANT -O3 -Wall -Wextra -Wpedantic -Wstrict-aliasing -Wcast-align -Wmissing-declarations -Wpointer-arith -Wcast-qual -Wzero-as-null-pointer-constant -Wnon-virtual-dtor -Ibuild/include -c src/tests/test_simd.cpp -o build/obj/test/test_simd.o
g++: error: unrecognized command line option ‘-mfpu=neon’
Makefile:326: recipe for target 'build/obj/test/test_simd.o' failed
make: *** [build/obj/test/test_simd.o] Error 1
```

The solution should be [this](https://stackoverflow.com/questions/29851128/gcc-arm64-aarch64-unrecognized-command-line-option-mfpu-neon).

**Comment by randombit** [comment]
I am confused. You indicate your GCC is configured for Aarch64. So, it is not possible to build for a standard 32-bit ARM target (which is what `arm32` is). Is your intention here to build for Aarch32?

**Comment by DevSidious** [comment]
My intention is to build for arm32 on an arm64 machine. The same way one would build Botan 32 bit on a 64 bit X64 machine.

It appears arm64 CPUs are able to run arm32 code, however Linuxes don't support mixed mode. I should be able to build though.

**Comment by randombit** [comment]
> I should be able to build though.

Execution works fine, but you need an arm32 compiler. In GCC at least, Aarch64 is a completely different backend from ARMv7 and company.

**Comment by DevSidious** [comment]
Well, the build goes on an on just fine until it reaches the line where `‘-mfpu=neon’` is passed to the compiler. According to the firs link, one shouldn't pass `‘-mfpu=neon’` to the compiler but rather `-O3` or `-O2 -ftree-vectorize` when building for arm32.

**Comment by randombit** [comment]
> Well, the build goes on just fine until it reaches the line where ‘-mfpu=neon’ is passed to the compiler. 

It is building, but is it building ARMv7 object files? I'm guessing not.

**Comment by DevSidious** [comment]
> It is building, but is it building ARMv7 object files? I'm guessing not.

I went and replaced `-mfpu=neon` with `-O2 -ftree-vectorize` on line 326 in the generated Makefile. Build completes successfully and tests complete too: `Tests complete ran 47596 tests in 119.11 sec all tests ok`.

This shows that the offending flag preventing a 32 bit build on AArch64 is indeed `-mfpu=neon`.

Currently waiting to see if the resulting binary (botan-test) actually runs on `armv7l`. My bet right now is that it won't. :D

**Comment by randombit** [comment]
You can test with `file`:
```
$ cat > hi.c              
int foo() { return 5; }
$ aarch64-linux-gnu-gcc -c hi.c -o hi.o
$ file hi.o 
hi.o: ELF 64-bit LSB relocatable, ARM aarch64, version 1 (SYSV), not stripped
$ arm-none-eabi-gcc -c hi.c -o hi.o
$ file hi.o
hi.o: ELF 32-bit LSB relocatable, ARM, EABI5 version 1 (SYSV), not stripped
```

**Comment by DevSidious** [comment]
```
root@dev-scw-xeni-a64:/tmp/botan# file botan-test
botan-test: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, BuildID[sha1]=73715262ee45a34952b4701d40438aa50d8445c3, not stripped
```

It seems I'm not getting a 32 bit executable. Of course it won't work on `armv7l`.

**Comment by DevSidious** [comment]
`./configure.py --cpu=arm32 --minimized-build --enable-modules=x509,ecdsa,rsa,emsa_pkcs1,sha2_64,aes --disable-shared-library --amalgamation --single-amalgamation-file`

... gives me a 64 bit build. Now I'm confused :| .

**Comment by randombit** [comment]
You'll need a compiler for 32-bit ARM, which will come in a completely different package from Aarch64. Eg on Ubuntu 16.04, gcc-5-aarch64-linux-gnu for 64-bit vs gcc-5-arm-linux-gnueabihf for 32-bit. It doesn't work like x86-32 vs x86-64, which share a backend in GCC and a simple -m32/-m64 flag to switch targets.

**Comment by randombit** [comment]
We have a compile-time check for x86-32 vs x86-64 cross-configuration because that's a common misconfiguration but for everwhere else, we just assume $CXX is going to produce binaries matching the CPU specified by `--cpu` without any verification. You compiled with an Aarch64 compiler, so that's the binaries you got, and it happened that ARMv7 is sufficiently like Aarch64 that - given the current source - it happened to work. But you could have said say `--cpu=mips64` and that would also work, unless you happened to hit a stray bit of inline asm or whatever and one or two files would fail to build.

We could/should have a better compile time check here, but fundamentally you need a different build of GCC for 32-bit ARM binaries.

**Comment by DevSidious** [comment]
Ok, I feel like I'm oozing with clarity right now. I have drawn some conclusions from my tests:

1. It seems that `configure.py` allows one to pass --cpu=arm32, resulting in a failure of the build, due to `-mfpu=neon`  compiler flags being passed for botan-test. This "protects" one from the scenario where one would expect a 32 bit build, but gets a 64 bit build instead. However, this is a clear case of bug as feature, because ARM7 toolchains do not understand `-mfpu=neon`  anyway, but rather `-O2 -ftree-vectorize` & friends.
2. If one tries to "correct" and replace `-mfpu=neon` with `-O2 -ftree-vectorize` while one does not have the ARM7 toolchain available, this will result in a successful build. However that would be the ""wrong" build, that is to say one would get a 64 bit build instead of a 32 bit build.

From 1. and 2. above it results that this scenario is a rich source of massive fuck-ups, inviting the question:

Can `configure.py` be updated to **slap one over the wrist really hard** in a scenario such as the one described in this thread?

**Comment by DevSidious** [comment]
Continuing to ooze with clarity, I could sum it up thus:

1. **Aarch64** does not need `-mfpu=neon` flag. It does not understand that, the same being achieved with `-O3` or `-O2 -ftree-vectorize`;
2. **ARM7** does need the  `-mfpu=neon` flag;
3. When invoking `configure.py` with `--cpu=arm32` on **Aarch64**, it presumes one would be using the **ARM7** tool-chain;
4. If one employs the **AArch64** tool-chain on **AArch64** instead of the **ARM7** tool-chain, `configure.py` won't have an issue with that. However, it adds `-mfpu=neon` which results in a build failure;
5. One tries to compensate for that by replacing `-mfpu=neon` with `-O2 -ftree-vectorize`. Build succeeds, however there is one caveat:
6. The resulting binaries are 64 bit AArch64, not 32 bit ARM7;

**Comment by randombit** [comment]
That pretty well sums it up. It's fairly easy to check the compiler's predefined macros to detect what actual arch it intends to build for which I've done in #1443. It errors out if it detects an inconsistency between what the user expressed and what the compiler seems to be doing.

**Comment by DevSidious** [comment]
It seems these changes don't pass CI. I'll test as soon as they do.

**Comment by noloader** [comment]
@DevSidious,

> My intention is to build for arm32 on an arm64 machine....

This is kind of unusual. What does `gcc -dumpmachine` have to say (for the compiler you are using)? That tells you what machine GCC can build for. It is called the ***host*** in Autotools speak.

As @randombit said, you can't build for arbitrary hosts. You can only build for the host the compiler is setup for.

> It appears arm64 CPUs are able to run arm32 code, however Linuxes don't support mixed mode. I should be able to build though.

This does not make much sense; neither does `--cpu=arm32` on Aarch64 unless you are cross-compiling. I don't recall seeing GCC options to generate 32-bit A-32 code for Aarch64, but I may have missed something. `-mabi=ilp32` looks close, but I don't see where you are using it. Also see [GCC's 3.18.1 AArch64 Options](https://gcc.gnu.org/onlinedocs/gcc/AArch64-Options.html).

You should probably ask some questions on the [GCC Help](https://gcc.gnu.org/lists.html) mailing list to see what you need to accomplish what you want to do. I believe you are interested in responses from a fellow named James Greenhalgh.

You might also want to talk with the Linaro folks. They may be able to help guide you with code generation needs. They have a GCC compiler mailing list. You are looking for responses from a fellow named Jim Wilson.

**Comment by DevSidious** [comment]
@noloader 

I apologize if I have not made my goals clear enough. The purpose of this exercise was to ascertain how would `configure.py` behave in the given context. Would it fail early or accept whatever and proceed ?

In this particular scenario, I did have the gcc-5-arm-linux-gnueabihf package installed, what I did not have was the aforementioned compiler incarnation set up as system default. 

The expectation was that `configure.py` would either successfully complete the configuration if all the prerequisites were in order **or** otherwise abort with a meaningful message.

The above expectation was not satisfied. `configure.py` accepted the `--cpu=arm32` parameter and proceeded to spit out a Makefile. The build failed, but it did so for questionable reasons: the  `-mfpu=neon` flag.

The fruits of this (albeit uncanny) scenario materialized in @randombit 's stance on the subject:

> We could/should have a better compile time check here

and

> #1443. It errors out if it detects an inconsistency between what the user expressed and what the compiler seems to be doing.

I wish to take the opportunity to thank both @randombit and @noloader for the information made available in this thread.
